pygdf/cudautils.py

```diff
@@ -378,33 +378,36 @@ def compute_scale(arr, vmin, vmax):
 
 
 @cuda.jit
-def gpu_label(arr, cat, val, out):
-    i = cuda.grid(1)
-    if i < out.size:
-        res = (val) if (arr[i] == cat) else arr[i]
-        out[i] = res
+def gpu_label(arr, labarr, out):
+    i, j = cuda.grid(1), cuda.grid(1)
+    for i in range(out.size):
```

Comment:
Should be looping on `labarr.shape[0]` instead.

Reply:
I would write the loop as:

```python

@cuda.jit
def gpu_label(arr, labarr, out):
    i = cuda.grid(1)

    if i < out.size:
        val = arr[i]
        out[i] = -1  # initialize for missing value
        for j in range(labarr.shape[0]):
            if val == labarr[j][0]:
                res = labarr[j][1]
                out[i] = res
                break  # found, stop loop
```

Reply:
`labarr.shape[0]` will have only unique cats, therefore its size will be less that `out.size`, it might not iterate for the whole array.

Reply:
The kernel is launched for all `out.size`.

```python
configured = gpu_label.forall(out.size)
configured(arr, labarr, out)
```

So, there is a hidden loop that move `i` from `0` to `out.size-1`.